add filestatus command

endorama opened the pull request:
Closes #460

Add `filestatus` command, reporting if the file is in encrypted or unencrypted state.

I reused `ensureNoMetadata` logic, thus the command would return `0` when the file is not encrypted and `203` when the file is encrypted (respecting the `codes.FileAlreadyEncrypted` error code)

It does not output the error message as returned by `ensureNoMetadata`, preferring a simpler output: "File is encrypted" or "File is unencrypted".

As I'm not so fond on `sops` internals, I'm sure there are multiple things to review, I'll gladly update the code to reflect any suggestion.

endorama commented on cmd/sops/main.go:
```diff
@@ -443,6 +444,55 @@ func main() {
+				// _, err := config.FindConfigFile(".")
+				// if err != nil {
+				// 	return common.NewExitError(err, codes.ErrorGeneric)
+				// }
```
Should this command check for the config file?

autrilla replied:
Not for now, at least, since we aren't using anything from the config file

endorama replied:
I'll remove this then.

endorama commented on cmd/sops/main.go:
```diff
@@ -443,6 +444,55 @@ func main() {
+					return common.NewExitError("File is encrypted", codes.FileAlreadyEncrypted)
+				}
+
+				fmt.Println("File is unencrypted")
```
Is `fmt` the proper way to output such an informative message?

autrilla replied:
Looks fine to me. I'm not sure whether we should use a machine-parseable and extensible format (probably JSON), though. It would let us put more status information there in the future, e.g. whether the MAC is okay.

endorama replied:
Would something like `{"encrypted": true|false}` be ok?

autrilla replied:
Looks good to me.

endorama commented on cmd/sops/main.go:
```diff
@@ -443,6 +444,55 @@ func main() {
+				}
+				if err := ensureNoMetadata(opts, branches[0]); err != nil {
+					return common.NewExitError("File is encrypted", codes.FileAlreadyEncrypted)
+				}
```
This code is a duplicate of https://github.com/mozilla/sops/blob/4b5b7ae308839fb4bbd6cc35851f93c01322643b/cmd/sops/encrypt.go#L58-L69

Should this duplication be avoided? If so, what would be the best way?

autrilla replied:
Yeah, I think we could extract that function to the `sops/cmd/common` package so that we can call it from both places.

autrilla replied:
Never mind, I misread -- no need to deduplicate this.

autrilla commented on cmd/sops/main.go:
```diff
@@ -443,6 +444,55 @@ func main() {
+			Usage:     "check the status of the file, returning encryption status",
+			ArgsUsage: `file`,
+			Flags:     []cli.Flag{},
+			Action: func(c *cli.Context) error {
```
Could you extract the body of this function to a package in the `subcommands` package? Like with https://github.com/mozilla/sops/blob/master/cmd/sops/subcommand/updatekeys/updatekeys.go and others. I think it makes the code easier to navigate vs having a lot of unrelated functionality in the same file.

endorama replied:
done in https://github.com/mozilla/sops/pull/545/commits/6e32720f45eb9a149f810540ee64e689d8662308

autrilla commented on cmd/sops/main.go:
```diff
@@ -443,6 +444,55 @@ func main() {
+					return common.NewExitError(fmt.Sprintf("Error unmarshalling file: %s", err), codes.CouldNotReadInputFile)
+				}
+				if err := ensureNoMetadata(opts, branches[0]); err != nil {
+					return common.NewExitError("File is encrypted", codes.FileAlreadyEncrypted)
```
Hmmm, I'm wondering whether we should return an error when the file is not encrypted, or the other way around. What was your reasoning for choosing this over the alternative?

endorama replied:
Mainly due to the fact that there was an error code for "file is already encrypted" and not the other way round. So I would expect users to be already used to such an error.

IMO none is an error, and using JSON output like you requested in a comment above everything makes more sense and this error can probably be removed.

autrilla replied:
Yeah, I agree. Let's make none of them be an error for now.

endorama commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+	return Status{Encrypted: hasMetadata}, nil
+}
+
+func checkMetadata(branch sops.TreeBranch) bool {
```
As `ensureNoMetadata` is declared private in `main` package (`cmd/sops/encrypt.go:48`) I duplicated the code, but `checkMetadata` could be in `common` package and being used in `ensureNoMetadata` too to avoid duplication.

autrilla replied:
Eh, this is simple enough that I think it's okay.

autrilla commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+	"go.mozilla.org/sops/cmd/sops/common"
+)
+
+// Opts rapresent input option for FileStatus
```
```suggestion
// Opts represent the input options for FileStatus
```

autrilla commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+	InputPath string
+	// KeyServices       []keyservice.KeyServiceClient
+	// KeyGroups         []sops.KeyGroup
+	// GroupThreshold    int
```
Remove the commented lines since they're not used

endorama replied:
Removed in commit https://github.com/mozilla/sops/pull/545/commits/80686aba224a6fed3aec410d0f75fc50065d2384

autrilla commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+	// GroupThreshold    int
+}
+
+// Status rapresent file status
```
```suggestion
// Status represents the status of a file
```

autrilla commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+
+// Status rapresent file status
+type Status struct {
+	Encrypted bool
```
```
// Encrypted represents whether the file provided is encrypted by SOPS
```

endorama replied:
Added in commit https://github.com/mozilla/sops/pull/545/commits/80686aba224a6fed3aec410d0f75fc50065d2384

endorama replied:
Resolved in 80686aba224a6fed3aec410d0f75fc50065d2384

jvehent commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+func checkMetadata(branch sops.TreeBranch) bool {
+	for _, b := range branch {
+		if b.Key == "sops" {
+			return true
```
isn't that a bit weak as a test? I'd suggest at least verifying that a `mac` is set and one key exists under the various sections.

autrilla replied:
It's indeed a bit weak. It's what we do elsewhere as an easy test to check whether the file is encrypted, but maybe as you say we should have a stricter test here.

endorama replied:
May I propose to go with this version for consistency, and work on a more complex test (that we can discuss more) on a separate one? I'm willing to work on that.
We can also extend a bit the scope and extract the check so both places can use the same code.

ajvb commented on cmd/sops/subcommand/filestatus/filestatus.go:
```diff
@@ -0,0 +1,47 @@
+	Encrypted bool
+}
+
+func FileStatus(opts Opts) (Status, error) {
```
Would be nice to have a function documentation comment for this function since it is public.

endorama replied:
Added in commit https://github.com/mozilla/sops/pull/545/commits/80686aba224a6fed3aec410d0f75fc50065d2384

kowalczykp commented:
Any update on that? It would be a really nice feature

mazdack commented:
any update? would be really nice to have this command

autrilla commented:
No updates. The review comments haven't been addressed in this PR, plus there's merge conflicts. We'd be happy to review a PR that builds on this one an handles the review comments.

endorama commented:
@autrilla May you let me know your opinion on https://github.com/mozilla/sops/pull/545#discussion_r339676101?

I was waiting on a reply there, other concerns should have been addressed as requested.

autrilla commented:
> @autrilla May you let me know your opinion on [#545 (comment)](https://github.com/mozilla/sops/pull/545#discussion_r339676101)?

It wouldn't take a lot of work to have a slightly better check by doing what jvehent suggests, e.g. ensuring there's a `mac` field.

endorama commented:
@autrilla I resolved conflicts and rebased onto latest `develop` so this should be now safe to merge again.

I started working on how to improve the check and I found out that using an implementation of `EncryptedFileLoader.LoadEncryptedFile(in []byte) (Tree, error)` on a non-encrypted file results in a `MetadataNotFound` error. Would this be ok as a deeper test? It feels "right" from my point of view. Is it enough or once metadata are found should it perform some other checks?

autrilla commented:
> @autrilla I resolved conflicts and rebased onto latest `develop` so this should be now safe to merge again.
> 
> I started working on how to improve the check and I found out that using an implementation of `EncryptedFileLoader.LoadEncryptedFile(in []byte) (Tree, error)` on a non-encrypted file results in a `MetadataNotFound` error. Would this be ok as a deeper test? It feels "right" from my point of view. Is it enough or once metadata are found should it perform some other check?

Trying to load the file sounds good to me.

onedr0p commented:
@autrilla any status here, or are you waiting on the author to rebase this PR again?

onedr0p commented:
Bump <3

autrilla commented:
Not all the issues are addressed. In particular, I asked for a better test that the file is encrypted. It's also missing documentation and tests, and as you mention, there's conflicts.

endorama commented:
Hello,

I have redone the implementation.   
The check is performed in the function `cfs` and:
- tries to load the encrypted file
- if it fails, report as not encrypted
- if encryption works, checks for `version` and `mac` keys to be present

`filestatus` subcommand now output JSON as requested (`{"encrypted": true|false}`).

I rebased onto master, so I'm not sure why it's still complaining about the conflict.

@autrilla is this implementation ok? I also added some test for it. The `cfs` function is private in the `filestatus` package, let me know where is the most appropriate place to put it into so it can be reused in other areas as well.

Thanks and sorry for the long wait!

Pehesi97 commented:
How is the progress in this PR? Do you need any help?

aDingil commented:
![alt text](https://media.tenor.com/Wyi2goAL_NcAAAAM/bump-reminder.gif)

ggordan commented:
Hi @ajvb @autrilla @jvehent, I see that roughly a year ago the milestone for this issue was changed to [v3.8.0](https://github.com/mozilla/sops/milestone/4).

Would that suggest that this PR is "complete" meaning that when v3.8.0 is released, this will be included, or are there still things that need to be addressed for the maintainers to be happy to merge this? If the answer is no, can we get a indication of what is needed, I'd be more than happy to help out.

Subsequently, is there a rough estimate on when v3.8.0 is expected to land?

Thanks,
Gordan

felixfontein commented:
@endorama would you mind rebasing another time and signing off your commits? Would be great if we could finally get this merged :)

davinkevin commented:
Can't wait to use it!

endorama commented:
Hello everybody and sorry for the long wait! I'd love to see this finally being merge and I have some time to dedicate to it.

Rebasing the old work was quite complex, so given the overall addition in this PR was small I re-implemented it on top of the latest `master` branch. I kept the implementation  the same with the same tests.

(The second force push is due to the DCO, which I didn't sign the first time)

Waiting on the review ❤

felixfontein commented:
I did some extensive tests, and it looks good! The method of detecting encrypted files is pretty good; obviously you can fake it with fake values, like the follwing YAML file:
```
sops:
    lastmodified: "2020-01-01T00:00:00Z"
    mac: x
    pgp:
        - created_at: "2020-01-01T00:00:00Z"
          enc: |-
            x
          fp: x
    version: x
```
(and similar variants for other storages), but I think that's acceptable, and it also seems to agree to what all reviewers so far asked for.

If nobody objects (especially from @getsops/maintainers side), I propose to merge this soon!

sabre1041 reviewed (approved):
LGTM

felixfontein commented:
@endorama thanks a lot for contributing this!

Also thanks a lot to everyone who reviewed this and commented on this over all these years! I'm glad this finally made it :)
